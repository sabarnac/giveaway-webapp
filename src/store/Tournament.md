Usage:

```tsx
import Config from "./store/config/Config";
import Round from "./store/round/Round";
import Participant from "./store/round/match/participant/Participant";

const tournament: Tournament = new Tournament(Config.instance);
const firstRound: Round = tournament.firstRound; // Round("round-1")
const lastRound: Round = tournament.lastRound; // Round("round-x")
const rounds: Round[] = tournament.rounds; // [Round("round-1"), ..., Round("round-x")]
const rounds: Participant = tournament.winner; // Participant("winner")
```
