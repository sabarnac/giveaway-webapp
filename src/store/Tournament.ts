import { observable, computed, action } from "mobx";
import Participant from "./Participant";
import last from "lodash.last";
import Config from "../config/Config";
import Round from "./Round";

/**
 * Class containing the details of a tournament.
 */
export default class Tournament {
  @observable private _rounds: Round[];
  private static _PARTICIPANTS: Participant[] = Config.allParticipants;

  private _createRound = (participants: Participant[]): Round =>
    new Round(participants);

  private _getRounds = (): Round[] => {
    const rounds: Round[] = [this._createRound(Tournament._PARTICIPANTS)];
    while (!this.hasWinner) {
      rounds.push(this._createRound(this._latestRound.winners));
    }
    return rounds;
  };

  public constructor() {
    this._rounds = this._getRounds();
  }

  @computed private get _latestRound(): Round {
    return <Round>last(this._rounds);
  }

  @computed private get hasWinner(): boolean {
    return this._latestRound.winners.length === 1;
  }

  /**
   * Gets the list of rounds in the tournament.
   * @returns {Round[]} The rounds in the tournament.
   */
  public get rounds(): Round[] {
    return this._rounds;
  }

  /**
   * Gets the final winner of the tournament.
   * @returns {Participant} The final winner.
   */
  @computed public get winner(): Participant {
    return <Participant>this._latestRound.winners[0];
  }
}
