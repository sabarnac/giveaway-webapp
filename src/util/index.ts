export let isDevEnvironment = (): boolean =>
  process.env.NODE_ENV === "development";

export let createAvatarImage = (identifier: string): string => {
  return `https://api.adorable.io/avatars/480/${identifier}`;
};

export let isInRange = (num: number, start: number, end: number): boolean =>
  num >= start && num <= end;
