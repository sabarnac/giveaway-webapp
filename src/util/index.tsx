import React, { ComponentType } from "react";
import { Dispatch, useState } from "react";
import Config from "../store/config/Config";
import { Redirect } from "react-router";
import { Observer } from "mobx-react";

/**
 * Returns if it is currently a dev environment.
 * @return {boolean} Whether it is currently a dev environment.
 */
export const isDevEnvironment = (): boolean =>
  process.env.NODE_ENV === "development";

/**
 * Returns a random avatar image URL.
 * @param {string} identifier The identifier to generate the avatar URL against.
 * @return {string} The avatar URL.
 */
export const createAvatarImage = (identifier: string): string => {
  return `https://api.adorable.io/avatars/480/${identifier}`;
};

/**
 * Checks if the number is within the range.
 * @param {number} num The number to check.
 * @param {number} start The start of the range.
 * @param {number} end The end of the range.
 * @return {boolean} Whether the number is within the range.
 */
export const isInRange = (num: number, start: number, end: number): boolean =>
  num >= start && num <= end;

/**
 * Type for custom animation state hook.
 */
export type AnimationStateHook = [
  number,
  () => void,
  (delay: number) => void,
  (state: number) => void
];

/**
 * Type for show overlay state hook.
 */
export type ShowOverlayHook = [boolean, (newValue: boolean) => void];

/**
 * Type for void function.
 */
export type VoidFunction = () => void;

/**
 * Runs the given action at least after the given delay.
 * @param {Function} action The action to run.
 * @param {number} delay The minimum delay to wait for.
 */
export const runOnDelay = (action: Function, delay: number): number =>
  setTimeout(action, delay);

/**
 * Create an animation state hook.
 * @return {AnimationStateHook} The current animation state, a state update method, and a delayed state update method.
 */
export const useAnimationState = (start: number = 0): AnimationStateHook => {
  const [currentState, setState]: [number, Dispatch<number>] = useState(start);

  const updateState = (): void => setState(currentState + 1);
  const updateStateDelay = (delay: number): VoidFunction => {
    const delayId: number = runOnDelay(updateState, delay);
    return () => clearTimeout(delayId);
  };

  return [currentState, updateState, updateStateDelay, setState];
};

/**
 * Normalize the given animation time by the animation speed multiplier.
 * @param {number} time The animation time to normalize.
 * @return {number} The normalized animation speed
 */
export const getNormalizedSpeed = (time: number): number =>
  time / Config.getInstance().speed;

/**
 * Runs the given action if the predicate is true.
 * @param {boolean} predicate The predicate to check.
 * @param {Function} action The action to run.
 */
export const runOnPredicate = (
  predicate: boolean,
  action: Function,
): VoidFunction | (() => VoidFunction) => (): void | VoidFunction =>
  predicate ? action() : undefined;

/**
 * Returns a redirect component for a given round and match.
 * @param {string} roundId The ID of the round to redirect to.
 * @param {string} matchId The ID of the match to redirect to.
 * @return {JSX.Element} The redirect element.
 */
export const getMatchRedirect = (roundId?: string, matchId?: string) => (
  <Redirect to={`/round/${roundId}/match/${matchId}`} />
);

/**
 * Returns a HOC with the given component wrapped by a MobX Observer component.
 * @param {ComponentType<T>} Component The component that should be under an observer.
 * @returns {(props: T) => JSX.Element} A HOC that wraps the given component under an observer.
 */
export const createObserver = <T extends {}>(
  Component: ComponentType<T>,
): ((props: T) => JSX.Element) => (props: T): JSX.Element => (
  <Observer>{() => <Component {...props} />}</Observer>
);
