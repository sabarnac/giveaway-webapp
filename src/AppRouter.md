Usage:

```tsx
import Config from "./store/config/Config";
import Tournament from "./store/Tournament";

const tournament: Tournament = new Tournament(Config.instance);
<AppRouter tournament={tournament} />;
```
