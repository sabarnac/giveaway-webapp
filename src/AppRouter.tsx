import { withRouter } from "react-router";
import AppRouter from "./AppRouter.react";

export default withRouter(AppRouter);
