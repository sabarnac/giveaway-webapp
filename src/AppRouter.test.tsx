import React from "react";
import AppRouter from "./AppRouter";
import Tournament from "./store/Tournament";
import { create, ReactTestRenderer } from "react-test-renderer";
import { createDummyTournament } from "./util/test";

describe("AppRouter Component.", (): void => {
  it("Matches snapshot.", (): void => {
    const tournament: Tournament = createDummyTournament();
    const component: ReactTestRenderer = create(
      <AppRouter tournament={tournament} />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
