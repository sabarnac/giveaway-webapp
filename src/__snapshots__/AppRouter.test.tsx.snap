// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppRouter Component. Matches snapshot. 1`] = `
<div>
  <div
    exact={true}
    path="/round/:roundId/match/:matchId"
    render={[Function]}
  />
  <div
    exact={true}
    path="/round/:roundId"
    render={[Function]}
  />
  <div
    exact={true}
    path="*"
    render={[Function]}
  />
</div>
`;
